cmd/snap-repair/runner.go

```diff
@@ -0,0 +1,188 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"gopkg.in/retry.v1"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/httputil"
+)
+
+// Runner implements fetching, tracking and running repairs.
+type Runner struct {
+	BaseURL *url.URL
+	cli     *http.Client
+}
+
+// NewRunner returns a Runner.
+func NewRunner() *Runner {
+	// TODO: pass TLSConfig with lower-bounded time
+	opts := httputil.ClientOpts{
+		Timeout:    15 * time.Second,
+		MayLogBody: false,
+	}
+	cli := httputil.NewHTTPClient(&opts)
+	return &Runner{
+		cli: cli,
+	}
+}
+
+var (
+	fetchRetryStrategy = retry.LimitCount(10, retry.LimitTime(1*time.Minute,
+		retry.Exponential{
+			Initial: 100 * time.Millisecond,
+			Factor:  2.5,
+		},
+	))
+
+	peekRetryStrategy = retry.LimitCount(7, retry.LimitTime(30*time.Second,
+		retry.Exponential{
+			Initial: 100 * time.Millisecond,
+			Factor:  2.5,
+		},
+	))
+)
+
+var ErrRepairNotFound = errors.New("repair not found")
+
+var (
+	maxRepairScriptSize = 24 * 1024 * 1024
+)
+
+// Fetch retrieves a stream with the repair with the given ids and any auxiliary assertions.
+func (run *Runner) Fetch(brandID, repairID string) (r []asserts.Assertion, err error) {
+	u, err := run.BaseURL.Parse(fmt.Sprintf("repairs/%s/%s", brandID, repairID))
```

Comment:
we could check that they have no "/"  in them if we wanted, that's a strong contraint on them anyway, it applies to anything that is used as part of the primary key of any assertion